src/renderer/wg_engine/tvgWgShaderTypes.cpp

```diff
@@ -138,34 +138,41 @@ void WgShaderTypeGradient::update(const RadialGradient* radialGradient)
 void WgShaderTypeGradient::updateTexData(const Fill::ColorStop* stops, uint32_t stopCnt)
 {
     assert(stops);
-    // head
+    // sort input data by offset
+    static Array<Fill::ColorStop> sstops(stopCnt);
+    sstops.clear();
+    sstops.push(stops[0]);
+    // filter by increasing offset
+    for (uint32_t i = 1; i < stopCnt; i++)
+        if (sstops.last().offset < stops[i].offset)
+            sstops.push(stops[i]);
     uint32_t range_s = 0;
-    uint32_t range_e = uint32_t(stops[0].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
+    uint32_t range_e = uint32_t(sstops[0].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
     for (uint32_t ti = range_s; (ti < range_e) && (ti < WG_TEXTURE_GRADIENT_SIZE); ti++) {
-        texData[ti * 4 + 0] = stops[0].r;
-        texData[ti * 4 + 1] = stops[0].g;
-        texData[ti * 4 + 2] = stops[0].b;
-        texData[ti * 4 + 3] = stops[0].a;
+        texData[ti * 4 + 0] = sstops[0].r;
+        texData[ti * 4 + 1] = sstops[0].g;
+        texData[ti * 4 + 2] = sstops[0].b;
+        texData[ti * 4 + 3] = sstops[0].a;
     }
     // body
-    for (uint32_t di = 1; di < stopCnt; di++) {
-        range_s = uint32_t(stops[di-1].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
-        range_e = uint32_t(stops[di-0].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
+    for (uint32_t di = 1; di < sstops.count; di++) {
+        range_s = uint32_t(sstops[di-1].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
+        range_e = uint32_t(sstops[di-0].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
```

Comment:
range_s = range_e;
        range_e = range_e = uint32_t(std::min(sstops[di].offset, 1.0f) * (WG_TEXTURE_GRADIENT_SIZE-1));
        for (uint32_t ti = range_s; ti < range_e; ti++) {